### DISCUSSION TITLE
Mutliscale Modeling

### DISCUSSION THREAD

**Discussion by manikumd22** [idaholab/moose#26549] [Q&A Modules: Solid mechanics]
Hi @recuero and @dschwen ,

I am working on multiscale modeling of materials, I am solving the microscale problem using machine learning, and for that, I have imported the weights and other parameters in the code.

For example, I am getting averaged stress from the micro problem as follows:

_S[qp] = W^T*_F[_qp]

So, How to relate this stress with the macro scale problem to be solve?

1. Do I need to obtain the material elastic constants as described in this paper: Development of a finite element based strain periodicity implementation
methodhttps://www.sciencedirect.com/science/article/pii/S0168874X20301165 

2. I was going through the ComputeNeoHookeanstress:

```
ComputeNeoHookeanstress::computeQpPK2Stress()
{

  // Large deformation = nonlinear strain
  if (_large_kinematics)
  {
    RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
    _S[_qp] = (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
    _C[_qp] =
        -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
        _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv);
  }
  // Small deformations = linear strain
  else
  {
    const auto I = RankTwoTensor::Identity();

    RankTwoTensor strain = 0.5 * (_F[_qp] + _F[_qp].transpose()) - I;
    _C[_qp] = _lambda[_qp] * I.times<i_, j_, k_, l_>(I) +
            2.0 * _mu[_qp] * RankFourTensor(RankFourTensor::initIdentitySymmetricFour);
    _S[_qp] = _C[_qp] * strain;

  }
}
```

Can I modify this material module and equate the last _S[qp] with my average stress from micro?

I'm confused how to link it

**Comment by recuero** [comment]
If I am understanding well, you just need a way to plug your local stress based on some parameters. 

> Can I modify this material module and equate the last _S[qp] with my average stress from micro?

In a way, yes. You need to understand what you need from MOOSE to obtain your local stress, i.e. `_S[qp]` and then create your stress material. In that case, your stress won't be NeoHookean, just the stress you compute from your microscale problem. One example of "micro" model plugged into the engineering scale simulations is the following object: 
 https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/LAROMANCEStressUpdateBase.C. But this object inherits from radial return, which may or may not be what you want.
